File: typescript-examples/src/module1/testDoubles/spy/__mocks__/EmailSenderSpy.ts
import {EmailSender} from "../EmailSender.ts";

export class EmailSenderSpy implements EmailSender {
    private sentEmails: Map<string, string> = new Map<string, string>()


    sendEmail(email: string, message: string): void {
        this.sentEmails.set(email, message);
    }

    emailSentTo(email: string) {
        return this.sentEmails.has(email);
    }
}

File: module-1-content/src/test/java/com/codurance/module1/TestDoubles/Spy/EmailSenderSpy.java
package com.codurance.module1.TestDoubles.Spy;

import java.util.ArrayList;
import java.util.List;

public class EmailSenderSpy implements EmailSender {
    private List<String> sentEmails = new ArrayList<>();

    @Override
    public void sendEmail(String recipient, String message) {
        sentEmails.add(recipient + ": " + message);
    }

    public boolean emailSentTo(String recipient) {
        return sentEmails.stream().anyMatch(email -> email.startsWith(recipient));
    }
}
